### DISCUSSION TITLE
Accessing and `flask.g` and `flask.current_app` in a `threading.Thread`

### DISCUSSION THREAD

**Discussion by lucpotage** [pallets/flask#5505] [Q&A]
I loop over a list and start a Thread. I don’t have access to `flask.app` and `flask.g` in the function executed by the Thread.

I tried to pass app as a parameter with no success. Any idea?

```.py
from flask import current_app as app

def my_function(app):
  with app.app_context():
        app.logger.info("Lorem ipsum")

def enqueue_my_function(queue, app):
    for chunk in my_function(app):
        queue.put(chunk)
    queue.put(None)

# ...
Thread(target=enqueue_my_function, args=(queue, app)).start()
```

I got this error:

> RuntimeError: Working outside of application context.
> 
> This typically means that you attempted to use functionality that needed the current application. To solve this, set up an application context with app.app_context(). See the documentation for more information.

I have read the documentation but still don’t understand the issue.

**Comment by ThiefMaster** [comment]
Pass `app._get_current_object()` instead of `app` to your Thread. The `current_app` proxy is thread-local so of course it won't work when accessed from a different thread.

And FWIW, aliasing `current_app as app` and then having a function arg named `app` as well is not helpful in making your code more readable (especially outside a small example snippet)

Also, the general advice: Depending on what you want to do, a background task daemon like Celery that runs in a different process may be better than running a thread within your webapp itself.

**Comment by lucpotage** [comment]
Thank you very much @ThiefMaster, it works!

```.py
from flask import current_app, g

def my_function(app):
  with app.app_context():
        app.logger.info("Lorem ipsum")
        g.something = "Dolor sit amet"

def enqueue_my_function(queue, app):
    for chunk in my_function(app):
        queue.put(chunk)
    queue.put(None)

# ...
Thread(target=enqueue_my_function, args=(queue, current_app. _get_current_object())).start()
```

I will have a look at background task daemon like Celery.
